Introduce `combined_fields` query

jtibshirani opened the pull request:
This PR introduces a new query called `combined_fields` for searching multiple
text fields. It takes a term-centric view, first analyzing the query string
into individual terms, then searching for each term any of the fields as though
they were one combined field. It is based on Lucene's `CombinedFieldQuery`,
which takes a principled approach to scoring based on the BM25F formula.

This query provides an alternative to the `cross_fields` `multi_match` mode. It
has simpler behavior and a more robust approach to scoring.

Addresses #41106.

jtibshirani committed:
Copy CombinedFieldQuery from Lucene 8.x.

jtibshirani committed:
Add new combined_fields query.

jtibshirani committed:
Move ZeroTermsQuery to its own class.

* Factor out query creation to this class.
* Rename to ZeroTermsQueryOption to clarify it isn't a query.

jtibshirani committed:
Make sure highlighting works.

jtibshirani committed:
Add documentation.

jtibshirani commented:
Some restrictions on the query (these are explained in more detail in the docs):
* All fields must be in the `text` type family. Additionally, all fields must have the same search analyzer.
* Currently only BM25 similarity is supported. Other default similarities or per-field similarities are not allowed.
* It has a simpler API than `multi_match` and omits several parameters, including `analyzer`, `fuzzines` and all related options, `lenient`, `slop`, and `tiebreaker`.

The PR is large because it also includes a refactor around `ZeroTermsQuery`. This is broken out into its own commit to make it easier to see what changed.

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

mayya-sharipova commented on docs/reference/query-dsl/combined-fields-query.asciidoc:
```diff
@@ -0,0 +1,179 @@
+accepts text fields that do not share the same analyzer.
+
+`multi_match` takes a field-centric view of the query by default. In contrast,
+`combined_fields` is term-centric: `operator` and `minimum_should_match` are
```
It would be nice to highlight how `combined_fields` is superior to `cross_fields` which is also term-centric.

jtibshirani replied:
No, `cross_fields` is also able to apply `operator` and `minimum_should_match` in a term-centric way. The main benefit of `combined_fields` is the robust and understandable scoring algorithm.

I tried to update this section to make it clearer:
* When mentioning field-centric scoring, make it clear that we're referring to `best_fields` and `most_fields`
* Add a sentence comparing to `cross_fields` and mentioning the benefit

mayya-sharipova commented on docs/reference/query-dsl/combined-fields-query.asciidoc:
```diff
@@ -0,0 +1,179 @@
+    +(combined("systems", fields:["title", "abstract"]))
+
+In other words, all terms must be present in at least one field for a
+document to match.
```
> all terms must be present in at least one field

May be a better way to phrase it as "each term much be present in at least one field"? 

Otherwise, it looks like at least one field much contain all terms, which is not true.

jtibshirani replied:
Good catch, 'each' is much clearer than 'all' here.

mayya-sharipova commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/360_combined_fields.yml:
```diff
@@ -0,0 +1,45 @@
+          body: "The concept of time is fundamental to our way of thinking..."
+
+  - do:
+      indices.refresh: {}
```
nit: `refresh: true` could be put as a part of previous request.

mayya-sharipova commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/360_combined_fields.yml:
```diff
@@ -0,0 +1,45 @@
+      reason: "combined fields query is not yet backported"
+  - do:
+      search:
+        rest_total_hits_as_int: true
```
As we want to deprecate `rest_total_hits_as_int` parameter, may be we don't need to introduce it in new tests.

jtibshirani replied:
Oops, this is a good point!

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/query/CombinedFieldsQueryBuilder.java:
```diff
@@ -0,0 +1,433 @@
+        if (Strings.isEmpty(field)) {
+            throw new IllegalArgumentException("supplied field is null or empty.");
+        }
+        checkNegativeBoost(boost);
```
do we also want check for  `boost < 1` similar to other places?

jtibshirani replied:
Yes I think we should! I had also forgotten to check in the `fields` method. I added a new test case for this, `CombinedFieldsQueryParsingTests#testInvalidFieldBoosts`.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/query/CombinedFieldsQueryBuilder.java:
```diff
@@ -0,0 +1,433 @@
+        for (int i = 0; i < size; i++) {
+            String field = in.readString();
+            float boost = in.readFloat();
+            checkNegativeBoost(boost);
```
do we need to check for `boost < 1` or considering that the stream came from encoding a valid `CombinedFieldsQueryBuilder` these checks are not necessary?

jtibshirani replied:
I copied this from `MultiMatchQueryBuilder` but I agree it doesn't seem necessary. Perhaps it was added there to handle the mixed-version case (where some older nodes could have negative boosts), but that doesn't apply to this new query.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/query/CombinedFieldsQueryBuilder.java:
```diff
@@ -0,0 +1,433 @@
+        }
+
+        @Override
+        protected Query analyzePhrase(String field, TokenStream stream, int slop) throws IOException {
```
Is this method ever called if we don't support phrases in a `combined_fields` query?

mayya-sharipova commented on server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java:
```diff
@@ -2113,6 +2115,45 @@ public void testMultiMatchQueryHighlight() throws IOException {
+            SearchResponse searchResponse = client().search(searchRequest("test").source(source)).actionGet();
+            assertHitCount(searchResponse, 1L);
+            assertHighlight(searchResponse, 0, "field1", 0,
+                anyOf(equalTo("<field1>The quick brown fox</field1> jumps over"),
```
I think it can never be `"<field1>The quick brown fox</field1> jumps over"`? No? Isn't this only applicable for phrase and phrase_prefix search?

jtibshirani replied:
Good point, I'll remove the first option here.

mayya-sharipova commented on server/src/test/java/org/elasticsearch/index/query/CombinedFieldsQueryParsingTests.java:
```diff
@@ -0,0 +1,355 @@
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+
+public class CombinedFieldsQueryParsingTests extends MapperServiceTestCase {
```
Great test coverage in these tests!!!

mayya-sharipova reviewed (approved):
@jtibshirani Thanks, this a great addition of a new query `combined_fields`.  I left a couple of comments with a main comment to make a little bit clear in documentation what is an advantage of `combined_fields` query over  cross_fields and other `multi_match`.
But overall this PR LGTM!

romseygeek commented on server/src/main/java/org/apache/lucene/search/XCombinedFieldQuery.java:
```diff
@@ -0,0 +1,466 @@
+/* @notice
```
I don't think this should be here?

jtibshirani replied:
Are you talking about the `@notice` annotation? This is required by our build for non-Elastic licensed code so that the notices are copied: #57017. I based this off other Lucene classes that we've copied like `XMoreLikeThis` and `IndexableBinaryStringTools`.

romseygeek commented on server/src/main/java/org/apache/lucene/search/XMultiNormsLeafSimScorer.java:
```diff
@@ -0,0 +1,161 @@
+/* @notice
```
And here?

romseygeek commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/CustomQueryScorer.java:
```diff
@@ -65,9 +65,7 @@ protected WeightedSpanTermExtractor newTermExtractor(String defaultField) {
-            if (terms.isEmpty()) {
-                extractWeightedTerms(terms, query, 1F);
-            }
+            extractWeightedTerms(terms, query, 1F);
```
Can you explain why this has changed?

jtibshirani replied:
This is tricky, I meant to call it out explicitly! Without this fix, the `plain` highlighter does not work on `combined_field` queries. Specifically, it will only highlight the first term, and fail to highlight subsequent terms because of the `terms.isEmpty()` check.

My understanding is that this was added in #15516 to work around a Lucene issue. The Lucene issue was later fixed: https://issues.apache.org/jira/browse/LUCENE-7112. All the tests added in #15516 still pass with this change.

romseygeek reviewed (commented):
Looks good to me overall.  I wonder if we need to implement `auto_generate_synonyms_phrase_query` and `zero_terms_query` for the new query?  Auto-generate in particular feels like a leftover from when we didn't really handle query-time token graphs and I think we should consider deprecating it elsewhere.

jtibshirani committed:
Improve test cases.

* Avoid using rest_total_hits_as_int
* Streamline check in HighlighterSearchIT

jtibshirani committed:
Correct field boost validation.

jtibshirani committed:
Improve documentation.

jtibshirani commented:
Thank you @mayya-sharipova and @romseygeek for the reviews! I pushed commits addressing your comments.

> with a main comment to make a little bit clear in documentation what is an advantage of combined_fields query over `cross_fields` and other `multi_match`.

I added some improvements to the docs section "Comparison to `multi_match` query". Let me know if this helps or if you think it's still unclear. For context, once this query is merged and users have tried it out, I plan to have a follow-up discussion about what to do with `cross_fields`. We may be able to deprecate and remove it, or maybe we'll identify some exact cases where it's helpful. After that discussion we can revise the docs to be even clearer.

> I wonder if we need to implement `auto_generate_synonyms_phrase_query` and `zero_terms_query` for the new query?

To me `zero_terms_query` seems helpful, we've seem users request it on query types where it was missing like `match_phrase` and `match_phrase_prefix`. I don't understand the details yet, but I'd be happy to remove `auto_generate_synonyms_phrase_query` if we don't think it's helpful anymore.

jtibshirani committed:
Merge remote-tracking branch 'upstream/master' into combined-fields-query

jtibshirani committed:
Fix documentation.

jtibshirani commented:
> I don't understand the details yet, but I'd be happy to remove `auto_generate_synonyms_phrase_query` if we don't think it's helpful anymore.

I caught up with @jimczi and now have a better understanding. @romseygeek I think you're suggesting just using the default of `true` and removing the possibility to disable it? I'd prefer to have a flag for now for consistency with other query types, which makes the API easier for users to understand. If we end up deprecating + removing it, we can just do it across all relevant query types, including `combined_fields`.

romseygeek reviewed (approved):
Works for me, thanks @jtibshirani

mayya-sharipova commented:
@jtibshirani  Thanks for explanations and new changes, they all LGTM.
